pravega/src/main/scala/akka/stream/alpakka/pravega/impl/PravegaTableReadFlow.scala

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.pravega.impl
+
+import java.util.concurrent.CompletableFuture
+
+import akka.annotation.InternalApi
+import akka.event.Logging
+import akka.stream.stage.{AsyncCallback, GraphStage, GraphStageLogic, InHandler, OutHandler, StageLogging}
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+
+import scala.util.control.NonFatal
+import scala.compat.java8.FutureConverters._
+import scala.concurrent.ExecutionContext.Implicits.global
+import akka.stream.alpakka.pravega.TableSettings
+
+import scala.util.{Failure, Try}
+import io.pravega.client.tables.KeyValueTable
+import io.pravega.client.KeyValueTableFactory
+
+import io.pravega.client.tables.KeyValueTableClientConfiguration
+
+import java.util.concurrent.atomic.AtomicInteger
+import io.pravega.client.tables.TableEntry
+import scala.util.Success
+@InternalApi private final class PravegaTableReadFlowStageLogic[A, B, K, V](
+    val shape: FlowShape[A, B],
+    val scope: String,
+    tableName: String,
+    tableSettings: TableSettings[K, V],
+    keyExtractor: A => Option[K],
+    combine: (A, Option[V]) => B,
+    familyExtractor: Option[A => String]
+) extends GraphStageLogic(shape)
+    with StageLogging {
+
+  private def in = shape.in
+  private def out = shape.out
+
+  private var keyValueTableFactory: KeyValueTableFactory = _
+  private var table: KeyValueTable[K, V] = _
+
+  private val onAir = new AtomicInteger
```

Comment:
Does this need to be an `AtomicInteger`? AFAICT this var isn't closed over in any non-stream-async code. By using the `AsyncCallback` you are ensuring that only the akka streams message processing thread will execute its body.

Reply:
Maybe rename to `inFlightEvents` or `inFlightMessages`.